README.md
R package "AssotesteR"
============================

This package is dedicated to statistical tests used in genetic association studies. It provides a toolbox with the state-of-the-art methods that have been developed to target the problem of testing for association between a binary phenotype and multiple genetic variants within a gene or a region. 

Installation options
-----------------------------
*  AssotesteR (stable version from CRAN)

   install.packages("AssotesteR")

*  AssotesteR (latest version from github)

   install.packages("devtools") 

   library(devtools)
   
   install_github('AssotesteR',  username='gastonstat')


Example Usage
-------------
    > library(AssotesteR)

    > cases = 500
    > controls = 500
    > total = cases + controls
    > phenotype = c(rep(1, cases), rep(0, controls))

    > set.seed(123)
    > genotype = matrix(rbinom(total*10, 2, 0.05), nrow=total, ncol=10)

    > mywss = WSS(phenotype, genotype, perm=500)


Or check out [www.gastonsanchez.com/assotester](http://www.gastonsanchez.com/assotester) for more information.

Links
-----
[AssotesteR package github](http://github.com/gastonstat/AssotesteR)

[The R Project](http://www.r-project.org/)

[gaston sanchez](http://www.gastonsanchez.com)


Author Contact
--------------
Gaston Sanchez (gaston.stat at gmail.com)


NEWS
Version 0.1-10 (2013-04-30)
------------------------------------------------------------
* `VT` changing permutation formula from '>' to '>='

Version 0.1-9 (2013-01-23)
------------------------------------------------------------
* `my_wss_method` changing order(score) to rank(score)

Version 0.1-8 (2013-01-22)
------------------------------------------------------------
* `WSS.R` changing perm.pval: x.perm '>=' wss.stat


R/VT.R
#'VT: Variable Threshold
#'
#'The Variable Threshold (VT) test has been proposed by Price et al (2010)
#'based on the assumption that the minor allele frequencies of the causal rare
#'variants may be different from those nonfunctional rare variants, which, if
#'true, can be utilized to improve the power of the corresponding pooled
#'association tests.  The idea behind this approach is that there exists some
#'(unknown) threshold T for which variants with a minor allele frequency (MAF)
#'below T are more likely to be functional than are variants with an MAF above
#'T. VT works by finding the maximum z-score across all possible values for the
#'threshold T.
#'
#'There is no imputation for the missing data. Missing values are simply
#'ignored in the computations.
#'
#'@param y numeric vector with phenotype status: 0=controls, 1=cases. No
#'missing data allowed
#'@param X numeric matrix or data frame with genotype data coded as 0, 1, 2.
#'Missing data is allowed
#'@param maf numeric value indicating the minor allele frequency threshold for
#'rare variants (must be a positive number between 0 and 1, \code{maf=0.05} by
#'default)
#'@param perm positive integer indicating the number of permutations (100 by
#'default)
#'@return An object of class \code{"assoctest"}, basically a list with the
#'following elements:
#'@returnItem vt.stat vt statistic
#'@returnItem perm.pval permuted p-value
#'@returnItem args descriptive information with number of controls, cases,
#'variants, and permutations
#'@returnItem name name of the statistic
#'@author Gaston Sanchez
#'@seealso \code{\link{WSS}}
#'@references Price AL, Kryukov GV, de Bakker PIW, Purcell SM, Staples J, Wei
#'LJ, Sunyaev SR (2010) Pooled Association Tests for Rare Variants in
#'Exon-Sequencing Studies. \emph{The American Journal of Human Genetics},
#'\bold{86}: 832-838
#'@examples
#'
#'  \dontrun{
#'  
#'  # number of cases
#'  cases = 500
#'
#'  # number of controls
#'  controls = 500
#'
#'  # total (cases + controls)
#'  total = cases + controls
#'
#'  #  phenotype vector
#'  phenotype = c(rep(1, cases), rep(0, controls))
#'
#'  # genotype matrix with 10 variants (random data)
#'  set.seed(1234)  
#'  genotype = matrix(rbinom(total*10, 2, 0.051), nrow=total, ncol=10)
#'
#'  # apply VT with maf=0.05 and 500 permutations
#'  myvt = VT(phenotype, genotype, maf=0.05, perm=500)
#'  myvt
#'  }
#'
VT <-
function(y, X, maf=0.05, perm=100)
{
  ## checking arguments
  Xy_perm = my_check(y, X, perm)
  y = Xy_perm$y
  X = Xy_perm$X
  perm = Xy_perm$perm
  if (mode(maf) != "numeric" || length(maf) != 1
      || maf <= 0  || maf >= 1)
    stop("argument 'maf' must be a value between 0 and 1")
  
  ## running vt method
  mafs = (1 + colSums(X, na.rm=TRUE)) / (2 + 2*nrow(X))
  h.maf = sort(unique(mafs))
  vt.stat = my_vt_method(y, X, mafs, h.maf)
  
  ## permutations
  perm.pval = NA
  if (perm > 0)
  {
    x.perm = rep(0, perm)
    for (i in 1:perm)
    {
      perm.sample = sample(1:length(y))
      x.perm[i] = my_vt_method(y[perm.sample], X, mafs, h.maf)
    }
    ## p-value
    perm.pval = sum(x.perm > vt.stat) / perm
  }
  
  ## results
  name = "VT: Variable Threshold"
  arg.spec = c(sum(y), length(y)-sum(y), ncol(X), maf, perm)
  names(arg.spec) = c("cases", "controls", "variants", "maf", "n.perms")	
  res = list(vt.stat = vt.stat, 
             perm.pval = perm.pval, 
             args = arg.spec, 
             name = name)
  class(res) = "assoctest"
  return(res)
}



src-x86_64/gdbr_dists.c
#include <R.h>
#include <Rmath.h>
#include <math.h>
/*
 *  gdbr_dists.c
 *  genomic similarity matrices  
 *
 */

void gdbr_IBS(int * Z, int * pn, int * pp, double * Sim)
{
    int i, j, k, n, p, diff;
    double temp;

    n = *pn;
    p = *pp;
    
    for(i=0; i<n-1; i++){
        for(j=i+1; j<n; j++){
            temp = 0;
            for(k=0; k<p; k++){
                diff = 2 - abs(Z[i*p + k] - Z[j*p + k]);
		        temp += diff ;
            }
            Sim[i*n + j] = Sim[j*n + i] = temp / 2/ p;
        }
    }

}


void gdbr_wIBS(int * Z, int * pn, int * pp, double * weight, double * Sim)
{
    int i, j, k, n, p, diff;
    double temp, w_total;

    n = *pn;
    p = *pp;
 
    w_total = 0;
    for(k=0; k<p; k++){
        w_total += weight[k];
    }

    for(i=0; i<n-1; i++){
        for(j=i+1; j<n; j++){
            temp = 0;
            for(k=0; k<p; k++){
                diff =  abs(Z[i*p + k] - Z[j*p + k]);
		        temp += diff*weight[k] ;
            }
            Sim[i*n + j] = Sim[j*n + i] = 1 - temp / 2 / w_total;
        }
    }
}




src-x86_64/skat_kernels.c
#include <R.h>
#include <Rmath.h>
#include <math.h>
/*
 *  skatkernels.c
 *
 */

void kernel_IBS(int * Z, int * pn, int * pp, double * Kernel)
{
    int i, j, k, n, p, diff;
    double temp;

    n = *pn;
    p = *pp;
    
    for(i=0; i<n-1; i++){
        for(j=i+1; j<n; j++){
            temp = 0;
            for(k=0; k<p; k++){
                diff = 2 - abs(Z[i*p + k] - Z[j*p + k]);
		        temp += diff ;
            }
            Kernel[i*n + j] = Kernel[j*n + i] = temp / 2/ p;
        }
    }
}


void kernel_wIBS(int * Z, int * pn, int * pp, double * weight, double * Kernel)
{
    int i, j, k, n, p, diff;
    double temp, w_total;

    n = *pn;
    p = *pp;
 
    w_total = 0;
    for(k=0; k<p; k++){
        w_total += weight[k];
    }

    for(i=0; i<n-1; i++){
        for(j=i+1; j<n; j++){
            temp = 0;
            for(k=0; k<p; k++){
                diff =  abs(Z[i*p + k] - Z[j*p + k]);
		        temp += diff*weight[k] ;
            }
            Kernel[i*n + j] = Kernel[j*n + i] = 1 - temp / 2 / w_total;
        }
    }
}


void kernel_twowayx(int * Z, int * pn, int * pp, double * Kernel)
{
    int i, j, k, n, p, diff;
    double temp, temp1, temp2;

    n = *pn;
    p = *pp;

    for(i=0; i<n; i++){
        for(j=i; j<n; j++){
            temp = 1;

            for(k=0; k<p; k++){
                temp2 = Z[i*p + k] * Z[j*p + k];
                temp += temp2;
                if(k==0){
                    temp1 = temp2;
                    continue ;
                }
                
                temp += temp1* Z[i*p + k] * Z[j*p + k];
                temp1 += temp2;

            }
            Kernel[i*n + j] = Kernel[j*n + i] = temp ;
        }
    }
}



R/WSS.R
#'WSS: Weighted Sum Statistic
#'
#'The WSS method has been proposed by Madsen and Browning (2009) as a pooling
#'approach. In WSS, rare-variant counts within the same gene for each
#'individual are accumulated rather than collapsing on them. Second, it
#'introduces a weighting term to emphasize alleles with a low frequency in
#'controls. Finally, the scores for all samples are ordered, and the WSS is
#'computed as the sum of ranks for cases. The significance is determined by a
#'permutation procedure.
#'
#'There is no imputation for the missing data. Missing values are simply
#'ignored in the computations.
#'
#'@param y numeric vector with phenotype status: 0=controls, 1=cases. No
#'missing data allowed
#'@param X numeric matrix or data frame with genotype data coded as 0, 1, 2.
#'Missing data is allowed
#'@param perm positive integer indicating the number of permutations (100 by
#'default)
#'@return An object of class \code{"assoctest"}, basically a list with the
#'following elements:
#'@returnItem wss.stat wss statistic
#'@returnItem perm.pval permuted p-value
#'@returnItem args descriptive information with number of controls, cases,
#'variants, and permutations
#'@returnItem name name of the statistic
#'@author Gaston Sanchez
#'@seealso \code{\link{CMC}}
#'@references Madsen BE, Browning SR (2009) A Groupwise Association Test for
#'Rare Mutations Using a Weighted Sum Statistic. \emph{PLoS Genetics},
#'\bold{5(2)}: e1000384
#'@examples
#'
#'  \dontrun{
#'  
#'  # number of cases
#'  cases = 500
#'
#'  # number of controls
#'  controls = 500
#'
#'  # total (cases + controls)
#'  total = cases + controls
#'
#'  # phenotype vector
#'  phenotype = c(rep(1, cases), rep(0, controls))
#'
#'  # genotype matrix with 10 variants (random data)
#'  set.seed(123)
#'  genotype = matrix(rbinom(total*10, 2, 0.05), nrow=total, ncol=10)
#'
#'  # apply WSS with 500 permutations
#'  mywss = WSS(phenotype, genotype, perm=500)
#'  mywss
#'  }
#'
WSS <-
function(y, X, perm=100)
{
  ## checking arguments
  Xy_perm = my_check(y, X, perm)
  y = Xy_perm$y
  X = Xy_perm$X
  perm = Xy_perm$perm
  
  ## running wss method
  wss.stat = my_wss_method(y, X)  
  
  ## permutations
  perm.pval = NA
  if (perm > 0)
  {
    x.perm = rep(0, perm)
    for (i in 1:perm)
    {
      perm.sample = sample(1:length(y))
      x.perm[i] = my_wss_method(y[perm.sample], X) 
    }
    # p-value 
    perm.pval = sum(x.perm > wss.stat) / perm
  }
  
  ## results
  name = "WSS: Weighted Sum Statistic"
  arg.spec = c(sum(y), length(y)-sum(y), ncol(X), perm)
  names(arg.spec) = c("cases", "controls", "variants", "n.perms")	
  res = list(wss.stat = wss.stat, 
             perm.pval = perm.pval, 
             args = arg.spec, 
             name = name)
  class(res) = "assoctest"
  return(res)
}

